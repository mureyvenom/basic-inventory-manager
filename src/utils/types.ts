import { ReactNode } from 'react';

export interface InventoryItem {
  name: string;
  stock: number;
  price: number;
  description: string;
  owner: string;
  date_created: string;
  last_modified: string;
}

export interface UserContextType {
  user: {
    email: string;
  };
  setUser: (e: string) => void;
}

export interface InventoryContextType {
  items: InventoryItem[] | null;
  saveItems: (items: InventoryItem[]) => void;
  addItem: (item: InventoryItem) => boolean;
  updateItem: (original: InventoryItem, replace: InventoryItem) => boolean;
  deleteItem: (item: InventoryItem) => void;
  getItems: () => InventoryItem[];
}

export interface SettingsContextType {
  darkmode: boolean;
  toggleDarkMode: () => void;
}

export interface ProviderProps {
  children: ReactNode;
}
