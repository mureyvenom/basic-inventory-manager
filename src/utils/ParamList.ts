import { InventoryItem } from './types';

export type MainStack = {
  Signin: undefined;
  InventoryList: undefined;
  Inventory: {
    mode: 'add' | 'edit';
    inventory?: InventoryItem;
  };
};
