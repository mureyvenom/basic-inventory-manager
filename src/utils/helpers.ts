export const numberWithCommas = (x: number, decimals?: number) => {
  return x.toFixed(decimals || 0).replace(/\B(?=(\d{3})+(?!\d))/g, ',');
};

export const addAlpha = (color: string, opacity: number) => {
  // coerce values so ti is between 0 and 1.
  var _opacity = Math.round(Math.min(Math.max(opacity || 1, 0), 1) * 255);
  return color + _opacity.toString(16).toUpperCase();
};

export const appConstants = () => {
  return {
    inventoryItemsKey: 'inventory-app-items',
    inventoryUserKey: 'inventory-app-user',
    inventorySettingsKey: 'inventory-app-settings',
  };
};

export const mockInventoryItems = [
  {
    date_created: '2023-03-11T02:27:30.776Z',
    description: 'Testing this out',
    last_modified: '2023-03-11T02:46:15.243Z',
    name: 'Testing this out',
    owner: 'holuwamurewa@gmail.com',
    price: 250,
    stock: 10,
  },
  {
    date_created: '2023-03-11T02:42:41.800Z',
    description: 'Second test',
    last_modified: '2023-03-11T02:48:05.826Z',
    name: 'Second test',
    owner: 'holuwamurewa@gmail.com',
    price: 1200,
    stock: 50,
  },
  {
    date_created: '2023-03-11T02:45:20.812Z',
    description: 'description',
    last_modified: '2023-03-11T02:45:20.812Z',
    name: 'Fourth entry',
    owner: 'holuwamurewa@gmail.com',
    price: 780,
    stock: 1000,
  },
];
