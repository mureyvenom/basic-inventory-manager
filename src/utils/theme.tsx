import { createTheme } from '@shopify/restyle';

const palette = {
  blue: '#6294CD',
  green: '#62CDA6',
  red: '#F92A4C',
  lightgreen: 'rgba(73,217,140,0.25)',
  lightred: 'rgba(255,102,127,0.25)',
  lightBg: 'rgba(255,255,255,0.05)',
  gray: '#2A3B3B',
  darkred: '#DA5757',
  darkgray: '#707070',
  black_: '#0A0A0A',
  lightBlue: '#29B9DC',
  purple: '#ED62F6',
  white: '#DDDDDD',
  muted: '#B0B0B0',
  faint: '#151515',
  transparent: 'transparent',
  constant_white: '#ffffff',
  constant_black: '#000000',
  // light theme colors
  bgLight: '#ffffff',
  faintLight: '#F9F9F9',
};

export const theme = createTheme({
  colors: {
    background: palette.black_,
    foreground: palette.white,
    primary: palette.blue,
    success: palette.green,
    danger: palette.red,
    lightsuccess: palette.lightgreen,
    lightdanger: palette.lightred,
    dangerdark: palette.darkred,
    secondary: palette.gray,
    info: palette.lightBlue,
    light: palette.purple,
    muted: palette.muted,
    muteddark: palette.darkgray,
    faint: palette.faint,
    white: palette.constant_white,
    black: palette.constant_black,
  },
  spacing: {
    xs: 4,
    s: 8,
    m: 16,
    l: 24,
    xl: 40,
    xxl: 80,
  },
  textVariants: {
    bold: {
      color: 'foreground',
      fontWeight: '700',
    },
    medium: {
      color: 'foreground',
      fontWeight: '600',
    },
    regular: {
      color: 'foreground',
    },
  },
  breakpoints: {},
});

export type Theme = typeof theme;

export const lightTheme: Theme = {
  ...theme,
  colors: {
    ...theme.colors,
    background: palette.bgLight,
    faint: palette.faintLight,
    foreground: palette.black_,
  },
};

export default theme;
