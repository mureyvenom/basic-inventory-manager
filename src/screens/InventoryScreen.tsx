import { RouteProp } from '@react-navigation/native';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import React, { useState } from 'react';
import { SafeAreaView, ScrollView } from 'react-native';
import Box from '../components/Box';
import { MainStack } from '../utils/ParamList';
import { useForm, Controller } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';
import { useUserContext } from '../state/providers/user.provider';
import Button from '../components/Button';
import { InventoryItem } from '../utils/types';
import { useInventoryContext } from '../state/providers/inventory.provider';
import Input from '../components/Input';
import ConfirmModal from '../components/ConfirmModal';
import Toast from 'react-native-toast-message';

const schema = yup
  .object({
    name: yup.string().required('Name is required'),
    description: yup
      .string()
      .required('Description is required')
      .min(3, 'Add a description of at least 3 characters'),
    price: yup.string().required('Price is required'),
    stock: yup.string().required('Stock count is required'),
  })
  .required();

interface Props {
  navigation: NativeStackNavigationProp<MainStack, 'Inventory'>;
  route: RouteProp<MainStack, 'Inventory'>;
}

const InventoryScreen = ({ navigation, route }: Props) => {
  const { user } = useUserContext();
  const { addItem, updateItem, deleteItem } = useInventoryContext();
  const [confirmDelete, setConfirmDelete] = useState(false);
  const { item } = route.params;

  const {
    control,
    handleSubmit,
    formState: { errors },
    // getValues,
  } = useForm({
    defaultValues: {
      name: item?.name || '',
      description: item?.description || '',
      price: item?.price || '0',
      stock: item?.stock || '0',
    },
    resolver: yupResolver(schema),
  });

  const submit = (cred: any) => {
    if (isNaN(parseInt(cred.stock)) || isNaN(Number(cred.price))) {
      Toast.show({
        type: 'error',
        text1: 'Error',
        text2: 'Stock and price must be valid numbers',
      });
      return;
    }
    if (route.params.mode === 'add') {
      const newItem: InventoryItem = {
        ...cred,
        date_created: new Date().toISOString(),
        last_modified: new Date().toISOString(),
        owner: user.email,
        price: Number(cred.price),
        stock: Number(cred.stock),
      };
      const response = addItem(newItem);
      if (response) {
        navigation.pop();
      } else {
        Toast.show({
          type: 'error',
          text1: 'Error',
          text2: 'Unable to add item',
        });
      }
    } else {
      const replace = {
        ...route.params?.item!,
        name: cred.name,
        description: cred.name,
        price: Number(cred.price),
        stock: Number(cred.stock),
        last_modified: new Date().toISOString(),
      };
      const response = updateItem(route.params?.item!, replace);
      if (response) {
        navigation.pop();
      } else {
        Toast.show({
          type: 'error',
          text1: 'Error',
          text2: 'Unable to edit item',
        });
      }
    }
  };

  return (
    <SafeAreaView style={{ flex: 1 }}>
      <ScrollView style={{ flex: 1 }}>
        <Box paddingTop="xxl" paddingHorizontal="l" paddingBottom="xxl">
          <Controller
            control={control}
            rules={{
              required: true,
            }}
            name="name"
            render={({ field: { onChange, value } }) => (
              <Input
                placeholder="Enter item name"
                value={value}
                onChangeText={v => onChange(v)}
                error={errors.name?.message}
                autoCapitalize="sentences"
                autoCorrect={false}
                label="Name"
              />
            )}
          />
          <Controller
            control={control}
            rules={{
              required: true,
            }}
            name="description"
            render={({ field: { onChange, value } }) => (
              <Input
                placeholder="Enter item description"
                value={value}
                onChangeText={v => onChange(v)}
                error={errors.description?.message}
                autoCapitalize="sentences"
                autoCorrect={false}
                label="Description"
              />
            )}
          />
          <Controller
            control={control}
            rules={{
              required: true,
            }}
            name="price"
            render={({ field: { onChange, value } }) => (
              <Input
                placeholder="Enter item price"
                value={value.toString()}
                onChangeText={v => onChange(v)}
                error={errors.price?.message}
                autoCapitalize="sentences"
                autoCorrect={false}
                keyboardType="number-pad"
                label="Price"
              />
            )}
          />
          <Controller
            control={control}
            rules={{
              required: true,
            }}
            name="stock"
            render={({ field: { onChange, value } }) => (
              <Input
                placeholder="Enter item stock"
                value={value.toString()}
                onChangeText={v => onChange(v)}
                error={errors.stock?.message}
                autoCapitalize="sentences"
                autoCorrect={false}
                keyboardType="number-pad"
                label="Stock"
              />
            )}
          />
          <Box marginBottom="l">
            <Button
              displayText={`${
                route.params.mode === 'add' ? 'Add' : 'Save'
              } Item`}
              onPress={handleSubmit(submit)}
            />
          </Box>
          {route.params.mode === 'edit' && (
            <Button
              displayText="Delete Item"
              buttonProps={{
                backgroundColor: 'danger',
              }}
              onPress={() => setConfirmDelete(true)}
            />
          )}
        </Box>
      </ScrollView>
      <ConfirmModal
        closeModal={() => setConfirmDelete(false)}
        confirmAction={() => {
          deleteItem(route.params.item!);
          navigation.pop();
        }}
        title="Are you sure you want to delete this item?"
        visible={confirmDelete}
      />
    </SafeAreaView>
  );
};

export default InventoryScreen;
