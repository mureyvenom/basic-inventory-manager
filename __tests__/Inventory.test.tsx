import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';
import { ThemeProvider } from '@shopify/restyle';
import theme from '../src/utils/theme';
import { render, fireEvent, act } from '@testing-library/react-native';
import { MainStack } from '../src/utils/ParamList';
import InventoryScreen from '../src/screens/InventoryScreen';
import { RouteProp } from '@react-navigation/native';
import { SettingsContext } from '../src/state/providers/settings.provider';
import { UserContext } from '../src/state/providers/user.provider';
import { InventoryContext } from '../src/state/providers/inventory.provider';
import { ProviderProps } from '../src/utils/types';
import { mockInventoryItems } from '../src/utils/helpers';

jest.useFakeTimers();

let navigation: NativeStackNavigationProp<MainStack, 'Inventory'>;
let route: RouteProp<MainStack, 'Inventory'>;

describe('Testing the inventory screen', () => {
  const RootProvider = ({ children }: ProviderProps) => {
    const darkmode = true;
    const toggleDarkMode = jest.fn();
    const addItem = jest.fn();
    const deleteItem = jest.fn();
    const getItems = () => mockInventoryItems;
    const saveItems = jest.fn();
    const updateItem = jest.fn();
    const setUser = (_: string) => {};

    return (
      <SettingsContext.Provider
        value={{
          darkmode,
          toggleDarkMode,
        }}>
        <UserContext.Provider
          value={{
            setUser,
            user: {
              email: 'holuwamurewa@gmail.com',
            },
          }}>
          <InventoryContext.Provider
            value={{
              addItem,
              deleteItem,
              getItems,
              saveItems,
              updateItem,
              items: mockInventoryItems,
            }}>
            {children}
          </InventoryContext.Provider>
        </UserContext.Provider>
      </SettingsContext.Provider>
    );
  };

  it('Create a snapshot of the screen', async () => {
    const tree = renderer
      .create(
        <RootProvider>
          <ThemeProvider theme={theme}>
            <InventoryScreen
              navigation={navigation}
              route={{
                ...route,
                params: {
                  mode: 'add',
                },
              }}
            />
          </ThemeProvider>
        </RootProvider>,
      )
      .toJSON();

    await renderer.act(async () => expect(tree).toMatchSnapshot());
  });

  it('Shows the confirmation modal when trying to delete item', async () => {
    const renderedComponent = render(
      <RootProvider>
        <ThemeProvider theme={theme}>
          <InventoryScreen
            navigation={navigation}
            route={{
              ...route,
              params: {
                mode: 'edit',
                item: mockInventoryItems[0],
              },
            }}
          />
        </ThemeProvider>
      </RootProvider>,
    );

    const deleteButton = renderedComponent.getByTestId('Delete');

    await act(async () => fireEvent.press(deleteButton));
    renderedComponent.getByTestId('confirmation-modal');
  });
});
