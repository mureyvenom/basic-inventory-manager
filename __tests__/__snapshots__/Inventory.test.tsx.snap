// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the inventory screen Create a snapshot of the screen 1`] = `
<RCTSafeAreaView
  style={
    {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    style={
      {
        "flex": 1,
      }
    }
  >
    <View>
      <View
        style={
          [
            {
              "paddingBottom": 80,
              "paddingHorizontal": 24,
              "paddingTop": 80,
            },
          ]
        }
      >
        <View
          style={
            [
              {
                "marginBottom": 24,
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#DDDDDD",
                  "marginBottom": 8,
                },
              ]
            }
          >
            Name
          </Text>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#0A0A0A",
                  "borderColor": "#B0B0B0",
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "height": 50,
                  "paddingHorizontal": 16,
                },
              ]
            }
          >
            <TextInput
              autoCapitalize="sentences"
              autoCorrect={false}
              onChangeText={[Function]}
              placeholder="Enter item name"
              placeholderTextColor="#B0B0B0"
              style={
                {
                  "color": "#DDDDDD",
                  "flex": 1,
                  "height": "100%",
                }
              }
              value=""
            />
          </View>
        </View>
        <View
          style={
            [
              {
                "marginBottom": 24,
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#DDDDDD",
                  "marginBottom": 8,
                },
              ]
            }
          >
            Description
          </Text>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#0A0A0A",
                  "borderColor": "#B0B0B0",
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "height": 50,
                  "paddingHorizontal": 16,
                },
              ]
            }
          >
            <TextInput
              autoCapitalize="sentences"
              autoCorrect={false}
              onChangeText={[Function]}
              placeholder="Enter item description"
              placeholderTextColor="#B0B0B0"
              style={
                {
                  "color": "#DDDDDD",
                  "flex": 1,
                  "height": "100%",
                }
              }
              value=""
            />
          </View>
        </View>
        <View
          style={
            [
              {
                "marginBottom": 24,
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#DDDDDD",
                  "marginBottom": 8,
                },
              ]
            }
          >
            Price
          </Text>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#0A0A0A",
                  "borderColor": "#B0B0B0",
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "height": 50,
                  "paddingHorizontal": 16,
                },
              ]
            }
          >
            <TextInput
              autoCapitalize="sentences"
              autoCorrect={false}
              keyboardType="number-pad"
              onChangeText={[Function]}
              placeholder="Enter item price"
              placeholderTextColor="#B0B0B0"
              style={
                {
                  "color": "#DDDDDD",
                  "flex": 1,
                  "height": "100%",
                }
              }
              value="0"
            />
          </View>
        </View>
        <View
          style={
            [
              {
                "marginBottom": 24,
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#DDDDDD",
                  "marginBottom": 8,
                },
              ]
            }
          >
            Stock
          </Text>
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#0A0A0A",
                  "borderColor": "#B0B0B0",
                  "borderRadius": 6,
                  "borderWidth": 1,
                  "flexDirection": "row",
                  "height": 50,
                  "paddingHorizontal": 16,
                },
              ]
            }
          >
            <TextInput
              autoCapitalize="sentences"
              autoCorrect={false}
              keyboardType="number-pad"
              onChangeText={[Function]}
              placeholder="Enter item stock"
              placeholderTextColor="#B0B0B0"
              style={
                {
                  "color": "#DDDDDD",
                  "flex": 1,
                  "height": "100%",
                }
              }
              value="0"
            />
          </View>
        </View>
        <View
          style={
            [
              {
                "marginBottom": 24,
              },
            ]
          }
        >
          <View
            style={
              [
                {
                  "width": "100%",
                },
              ]
            }
          >
            <View
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": undefined,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "backgroundColor": "#6294CD",
                      "borderRadius": 6,
                      "flexDirection": "row",
                      "height": 50,
                      "justifyContent": "center",
                      "paddingHorizontal": 24,
                    },
                    {},
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "#ffffff",
                        "fontSize": 16,
                        "fontWeight": "600",
                      },
                    ]
                  }
                >
                  Add Item
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <Modal
    animationType="none"
    deviceHeight={null}
    deviceWidth={null}
    hardwareAccelerated={false}
    hideModalContentWhileAnimating={false}
    onBackdropPress={[Function]}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onRequestClose={[Function]}
    panResponderThreshold={4}
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    statusBarTranslucent={false}
    supportedOrientations={
      [
        "portrait",
        "landscape",
      ]
    }
    swipeThreshold={100}
    testID="confirmation-modal"
    transparent={true}
    visible={false}
  />
</RCTSafeAreaView>
`;
